### PR TITLE
feat(TG-289): Constructor - Dashboard view creating- NEW SCOPE

### PR DESCRIPTION
**Pull request by valakar**

### CODE REVIEW

**Comment by review-ninja** [comment]
<a href="https://app.review.ninja/garage-it/SmartHouse-frontend/pull/171" target="_blank">ReviewNinja

File: src/app/constructor/dashboard/dashboard-constructor.component.test.ts [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+                }
+            }
+        };
+        sut = new DashboardConstructorComponent(ActivatedRoute);
+    });
+
+    describe('on init', () => {
+        describe('when dashboard view is empty', () => {
+            it('should receive devices from state', () => {
+                sut.ngOnInit();
```
Comment (velmyk) [inline src/app/constructor/dashboard/dashboard-constructor.component.test.ts:27]: You can move it to beforeEach section

File: src/app/constructor/dashboard/dashboard-constructor.component.test.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+                expect(sut.widgets).toEqual([]);
+            });
+        });
+
+        describe('when dashbord has some widgets', () => {
+            beforeEach(() => {
+                ActivatedRoute.snapshot.data.widgets = mockWidgets;
+            });
+            it('should receive devices from state that is not active widgets', () => {
+                sut.ngOnInit();
```
Comment (velmyk) [inline src/app/constructor/dashboard/dashboard-constructor.component.test.ts]: You can move it to beforeEach section

File: src/app/constructor/dashboard/dashboard-constructor.component.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+    constructor(private route: ActivatedRoute) {
+    }
+
+    ngOnInit(): void {
+        this.widgets = this.route.snapshot.data['widgets'];
+        const widgetsIds = this.widgets.map(widget => widget.mqttId);
+
+        this.sensors = this.route.snapshot.data['sensors']
+            .filter(device => !(widgetsIds.indexOf(device.mqttId) + 1));
```
Comment (velmyk) [inline src/app/constructor/dashboard/dashboard-constructor.component.ts:23]: You can use .includes method here

Comment (yakimko) [reply src/app/constructor/dashboard/dashboard-constructor.component.ts:23]: typescript doesn't have includes method for arrays

```
.filter(device => widgetsIds.indexOf(device.mqttId) === -1);
```

File: src/app/dashboard/shared/base-output-sensor/base-output-sensor.test.ts [hunk 4]
```diff
@@ -137,23 +137,46 @@ describe('base-output-sensor', () => {
...
+                beforeEach(() => {
+                    jasmine.clock().install();
+                    jasmine.clock().mockDate(currentDate);
+
+                    sut[pending] = null;
+                    sut.onDeviceDataChanged(data);
+                });
+
+                it('should convert data from device`s format using function', () => {
+                    expect(sut.fromDeviceRepresentation)
```
Comment (velmyk) [inline src/app/dashboard/shared/base-output-sensor/base-output-sensor.test.ts:167]: Check one assertion per test

File: src/app/dashboard/shared/base-output-sensor/base-output-sensor.test.ts [hunk 5]
```diff
@@ -137,23 +137,46 @@ describe('base-output-sensor', () => {
...
+
+                it('should convert data from device`s format', () => {
+                    expect(sut.data.value).toEqual(convertedValue);
+                });
+
+                it('should update last update time', () => {
+                    expect(sut.data.updateTime).toEqual(currentDate);
+                });
+
+                afterEach(() => {
```
Comment (velmyk) [inline src/app/dashboard/shared/base-output-sensor/base-output-sensor.test.ts]: Better to declare this on top of test near beforeEach

File: src/app/dashboard/shared/base-sensor/base-sensor.test.ts [hunk 6]
```diff
@@ -53,6 +66,32 @@ describe('base-sensor', () => {
...
+                mqttId: deviceId
+            };
+            sut.onDeviceDataChanged(data);
+        });
+
+        it('should update last update time', () => {
+            expect(sut.data.updateTime).toEqual(currentDate);
+        });
+
+        afterEach(() => {
```
Comment (velmyk) [inline src/app/dashboard/shared/base-sensor/base-sensor.test.ts:86]: Declare on top

File: src/app/constructor/dashboard/dashboard-constructor.component.test.ts (hunk 1 above)
Comment (valakar) [inline src/app/constructor/dashboard/dashboard-constructor.component.test.ts:27]: !fixed

File: src/app/constructor/dashboard/dashboard-constructor.component.test.ts (hunk 2 above)
Comment (valakar) [reply src/app/constructor/dashboard/dashboard-constructor.component.test.ts]: !fixed

File: src/app/dashboard/shared/base-output-sensor/base-output-sensor.test.ts (hunk 5 above)
Comment (valakar) [reply src/app/dashboard/shared/base-output-sensor/base-output-sensor.test.ts]: !fixed

File: src/app/dashboard/shared/base-sensor/base-sensor.test.ts (hunk 6 above)
Comment (valakar) [reply src/app/dashboard/shared/base-sensor/base-sensor.test.ts:86]: !fixed

File: src/app/constructor/map/map-constructor.component.ts [hunk 7]
```diff
@@ -1,7 +1,22 @@
...
+    private sensors: Device[] = [];
+
+    constructor(private route: ActivatedRoute) {
+    }
+
+    ngOnInit(): void {
+        this.sensors = this.route.snapshot.data['sensors'];
+    }
+
+    onAddSensor(sensor): void {
```
Comment (piton13) [inline src/app/constructor/map/map-constructor.component.ts:20]: Empty method should be removed.

File: src/app/constructor/map/map-constructor.component.ts [hunk 8]
```diff
@@ -1,7 +1,22 @@
 import { Component } from '@angular/core';
+import { ActivatedRoute } from '@angular/router';
+import { Device } from '../../devices/device.model';
 
 @Component({
     selector: 'sh-map-constructor',
     templateUrl: './map-constructor.template.html'
 })
-export class MapConstructorComponent {}
+export class MapConstructorComponent {
```
Comment (piton13) [inline src/app/constructor/map/map-constructor.component.ts:9]: This class should implement OnInit interface.

File: src/app/dashboard/sensor-executor-widget/sensor-executor-widget.component.test.ts [hunk 9]
```diff
@@ -1,4 +1,5 @@
 import { SensorExecutorWidgetComponent } from './sensor-executor-widget.component';
+import { EventEmitter } from '@angular/core';
```
Comment (piton13) [inline src/app/dashboard/sensor-executor-widget/sensor-executor-widget.component.test.ts:2]: If we use the unit testing approach then we should refuse of importing additional services and components.

File: src/app/dashboard/sensor-executor-widget/sensor-executor-widget.template.html [hunk 10]
```diff
@@ -1,4 +1,14 @@
 <div class="sensor-widget">
+    <button class="sensor-widget_button"
+            (click)="removeWidget()">
+        x
+    </button>
+
+    <span class="sensor-widget_time"
+          [hidden]="data.updateTime===undefined">
```
Comment (piton13) [inline src/app/dashboard/sensor-executor-widget/sensor-executor-widget.template.html:8]: Maybe will be better to use '!!data.updateTime'?

File: src/app/dashboard/sensor-regular-widget/sensor-regular-widget.component.test.ts [hunk 11]
```diff
@@ -0,0 +1,14 @@
+import { SensorRegularWidgetComponent } from './sensor-regular-widget.component';
+import { EventEmitter } from '@angular/core';
```
Comment (piton13) [inline src/app/dashboard/sensor-regular-widget/sensor-regular-widget.component.test.ts:2]: the same.

File: src/app/dashboard/sensor-servo-widget/sensor-servo-widget.template.html [hunk 12]
```diff
@@ -1,4 +1,13 @@
 <div class="sensor-widget sensor-servo-widget">
+    <button class="sensor-widget_button"
+            (click)="removeWidget()">
+        x
+    </button>
+
+    <span class="sensor-widget_time"
+          [hidden]="data.updateTime===undefined">
```
Comment (piton13) [inline src/app/dashboard/sensor-servo-widget/sensor-servo-widget.template.html:8]: the same

File: src/app/dashboard/sensor-widget/sensor-widget.component.test.ts [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+        expect(sut.onRemoveWidget instanceof EventEmitter).toBeTruthy();
+    });
+
+    describe('when remove widget', () => {
+        beforeEach(() => {
+            sut.removeWidget();
+        });
+
+        it('should emit event', () => {
+            expect(sut.onRemoveWidget.emit).toHaveBeenCalled();
```
Comment (piton13) [inline src/app/dashboard/sensor-widget/sensor-widget.component.test.ts:22]: Will be better to test with which value the method was called.

File: src/app/constructor/map/map-constructor.component.ts (hunk 7 above)
Comment (valakar) [reply src/app/constructor/map/map-constructor.component.ts:20]: will be implemented in creating map-view story

File: src/app/dashboard/sensor-executor-widget/sensor-executor-widget.component.test.ts (hunk 9 above)
Comment (valakar) [reply src/app/dashboard/sensor-executor-widget/sensor-executor-widget.component.test.ts:2]: I have to test that onRemoveWidget is real EventEmitter

File: src/app/dashboard/sensor-executor-widget/sensor-executor-widget.template.html (hunk 10 above)
Comment (valakar) [inline src/app/dashboard/sensor-executor-widget/sensor-executor-widget.template.html:8]: It don't want to work in such way

**Pull request merged**